CONTRIBUTING.md

```diff
@@ -0,0 +1,289 @@
+
+**Note** that the tests in `cardano-ledger-shelley-test` require two Ruby gems,
+[cbor-diag](https://rubygems.org/gems/cbor-diag) and
+[cddl](https://rubygems.org/gems/cddl).
+
+It can be helpful to use the `--test-show-details=streaming` option for seeing
+the output of the tests while they run:
+
+```shell
+cabal test cardano-ledger-shelley-test --test-show-details=streaming
+```
+
+### Running Specific Tests
+
+The test suites use [Tasty](https://github.com/feuerbach/tasty),
+which allows for running specific tests.
+This is done by passing the `-p` flag to the test program, followed by an `awk` pattern.
+You can alternatively use the `TASTY_PATTERN` environment variable with a pattern.
+For example, the Shelley golden tests can be run with:
+
+```shell
+cabal test cardano-ledger-shelley-test --test-options="-p golden"
+```
+
+or
+
+```shell
+TASTY_PATTERN=golden cabal test cardano-ledger-shelley-test
+```
+
+`Tasty` allows for more
+[complex patterns](https://github.com/feuerbach/tasty#patterns).
+For instance, to run only the Byron update mechanism tests for the ledger
+that classify traces, we can pass the
+`-p $1 ~ /Ledger/ && $2 ~ /Update/ && $3 ~ /classified/` option.
+Here each `$i` refers to a level in the tests names hierarchy.
+Passing `-l` to `tasty` will list the available test names.
+
+When testing using `cabal`, pay special attention to escaping the right symbols, e.g.:
+
+```shell
+cabal test byron-spec-ledger:test:byron-spec-ledger-test --test-options "-p \"\$1 ~ /Ledger/ && \$2 ~ /Update/ && \$3 ~ /classified/\""
+```
+
+### Replaying QuickCheck Failures
+
+When a QuickCheck test fails, the seed which produced the failure is reported.
+The failure can be replayed with:
+
+```shell
+cabal test cardano-ledger-shelley-test --test-options "--quickcheck-replay=42"
+```
+(where 42 is an example seed).
+
+### Test Scenarios
+
+Most of the test suites are grouped into test scenarios.
+For example, the Shelley test suite contains
+`ContinuousIntegration`, `Development`, `Nightly`, and `Fast`,
+which can be run with the `--scenario` flag. For example:
+
+```shell
+cabal test cardano-ledger-shelley-test --test-options --scenario=Fast
+```
+
+### ghcid
+
+We have support for running
+[ghcid](https://github.com/ndmitchell/ghcid)
+from inside of nix-shell.
+Enter nix-shell from the base directory of the repository,
+change directories to the cabal package that you wish to check,
+then run `ghcid`.
+
+For example:
+
+```shell
+nix-shell
+cd eras/shelley/impl/
+ghcid
+```
+
+## nix-build Infrastructure
+
+The artifacts in this repository can be built and tested using nix. This is
+additionally used by the Hydra CI to test building, including cross-compilation
+for other systems.
+
+### To add a new Haskell project
+
+To add a new Haskell project, you should do the following:
+
+1. Create the project in the usual way. It should have an appropriate `.cabal` file.
+2. Add the project to the [top-level stack.yaml](./stack.yaml), configuring
```

Comment:
This is out of date, we don't have a stack.yaml

Reply:
great catch, thanks! it looks like nearly that entire section on adding a new Haskell project was out of date. I made my edit in a separate commit so you could see what small part of it I salvaged.